src/passes/ConstantFieldPropagation.cpp

```diff
@@ -415,14 +427,27 @@ struct PCVScanner
     info.note(Literal::makeZero(fieldType));
   }
 
-  void noteCopy(HeapType type, Index index, PossibleConstantValues& info) {
+  void noteCopy(HeapType type,
+                Index index,
+                MemoryOrder order,
+                PossibleConstantValues& info) {
     // Note copies, as they must be considered later. See the comment on the
     // propagation of values below.
     functionCopyInfos[getFunction()][type][index] = true;
   }
 
-  void noteRead(HeapType type, Index index, PossibleConstantValues& info) {
-    // Reads do not interest us.
+  void noteRead(HeapType type,
+                Index index,
+                MemoryOrder order,
+                PossibleConstantValues& info) {
+    // Reads do not interest us, except that acquire reads should prohibit
+    // optimization. If we optimized fields with acquire reads, we would have to
+    // replace the gets with acquire fences, which are potentially more
+    // expensive than the original acquire reads because acquire fences are
+    // stronger than acquire reads.
```

Comment:
Can't we apply this reasoning on line 177? (if acqrel, do not optimize the StructGet there, instead of the new assert)

Reply:
Hmm good question. The current implementation inhibits all optimization of fields that are ever accessed with an acqrel get, but we could be more precise and inhibit only the optimization of the specific acqrel gets and still optimize non-acqrel gets of the field. I'll do that.

Reply:
This was a huge simplification. PTAL!